File: app/src/main/java/com/appliedrec/credentials/app/MainActivity.java
package com.appliedrec.credentials.app;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Rect;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.util.Pair;
import androidx.exifinterface.media.ExifInterface;

import com.appliedrec.barcodedatamatcher.DocumentFrontPageData;
import com.appliedrec.credentials.app.databinding.ActivityMainBinding;
import com.appliedrec.verid.core2.Classifier;
import com.appliedrec.verid.core2.Face;
import com.appliedrec.verid.core2.FaceDetection;
import com.appliedrec.verid.core2.FaceDetectionImage;
import com.appliedrec.verid.core2.IFaceDetection;
import com.appliedrec.verid.core2.RecognizableFace;
import com.appliedrec.verid.core2.VerIDImageBitmap;
import com.microblink.entities.recognizers.Recognizer;
import com.microblink.entities.recognizers.RecognizerBundle;
import com.microblink.entities.recognizers.blinkid.DataMatchResult;
import com.microblink.entities.recognizers.blinkid.generic.BlinkIdCombinedRecognizer;
import com.microblink.entities.recognizers.blinkid.generic.viz.VizResult;
import com.microblink.uisettings.ActivityRunner;
import com.microblink.uisettings.BlinkIdUISettings;

import java.util.Date;
import java.util.GregorianCalendar;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class MainActivity extends BaseActivity {

    private static final int REQUEST_CODE_SCAN_ID_CARD = 1;
    private RecognizerBundle recognizerBundle;
    private ActivityMainBinding viewBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());
        viewBinding.button.setOnClickListener(v -> scanIDCard());
        viewBinding.progressBar.setVisibility(View.VISIBLE);
        viewBinding.mainUI.setVisibility(View.INVISIBLE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        viewBinding = null;
    }

    @Override
    public void onVerIDPropertiesAvailable() {
        BlinkLicenceKeyUpdater licenceKeyUpdater = new BlinkLicenceKeyUpdater(this);
        addDisposable(licenceKeyUpdater
                .getSavedLicenceKey()
                .flatMapCompletable(licenceKeyUpdater::setLicenceKey)
                .onErrorResumeNext(error -> licenceKeyUpdater.deleteSavedLicenceKey()
                        .andThen(licenceKeyUpdater.getSavedLicenceKey())
                        .flatMapCompletable(licenceKeyUpdater::setLicenceKey))
                .onErrorResumeNext(error -> licenceKeyUpdater
                        .getLicenceKeyFromRemote()
                        .flatMapCompletable(licenceKeyUpdater::setLicenceKey))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        () -> {
                            viewBinding.progressBar.setVisibility(View.GONE);
                            viewBinding.mainUI.setVisibility(View.VISIBLE);
                        },
                        error -> {
                            viewBinding.progressBar.setVisibility(View.GONE);
                            new AlertDialog.Builder(this)
                                    .setTitle(R.string.invalid_mb_key)
                                    .setPositiveButton(android.R.string.ok, (dialogInterface, i) -> finish())
                                    .setCancelable(false)
                                    .create()
                                    .show();
                        }
                ));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.about) {
            showAbout();
        } else if (item.getItemId() == R.id.settings) {
            showSettings();
        } else if (item.getItemId() == R.id.supported_documents) {
            showSupportedDocuments();
        } else {
            return super.onOptionsItemSelected(item);
        }
        return true;
    }

    private void showCardFromResult(BlinkIdCombinedRecognizer.Result result) {
        byte[] frontImage = result.getEncodedFrontFullDocumentImage();
        @SuppressLint("CheckResult")
        Single<Pair<FaceWithImage,Float>> single = Single.<Bitmap>create(emitter -> {
            try {
                if (frontImage.length == 0) {
                    throw new Exception(getString(R.string.failed_to_collect_card_image));
                }
                Bitmap frontImageBitmap = BitmapFactory.decodeByteArray(frontImage, 0, frontImage.length);
                emitter.onSuccess(frontImageBitmap);
            } catch (Exception e) {
                emitter.onError(e);
            }
        }).flatMap(bitmap -> emitter -> {
            try {
                VerIDImageBitmap image = new VerIDImageBitmap(bitmap, ExifInterface.ORIENTATION_NORMAL);
                Face[] faces = getVerID().getFaceDetection().detectFacesInImage(image.createFaceDetectionImage(), 1, 0);
                if (faces.length > 0) {
                    RecognizableFace[] recognizableFaces = getVerID().getFaceRecognition().createRecognizableFacesFromFaces(faces, image);
                    Float authenticityScore = null;
                    try {
                        if (AuthenticityScoreSupport.defaultInstance().isDocumentSupported(result)) {
                            IFaceDetection<FaceDetectionImage> faceDetection = getVerID().getFaceDetection();
                            FaceDetection veridFaceDetection = (FaceDetection)faceDetection;
                            Classifier[] authenticityClassifiers = AuthenticityScoreSupport.defaultInstance().getClassifiers(this);
                            for (Classifier classifier : authenticityClassifiers) {
                                float score = veridFaceDetection.extractAttributeFromFace(faces[0], image, classifier.getName());
                                if (authenticityScore == null || score > authenticityScore) {
                                    authenticityScore = score;
                                }
                            }
                        }
                    } catch (Exception ignore) {}
                    emitter.onSuccess(new FaceWithImage(recognizableFaces[0], bitmap, authenticityScore));
                } else {
                    throw new Exception(getString(R.string.failed_to_detect_face_in_card));
                }
            } catch (Exception e) {
                emitter.onError(e);
            }
        }).zipWith(new FrontBackMatcher().getFrontBackMatchScore(result).onErrorReturn(error -> null), (faceWithImage, frontBackMatchScore) -> new Pair<>((FaceWithImage) faceWithImage, frontBackMatchScore));
        addDisposable(single.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(
                faceCapture -> showCard(faceCapture.first, new DocumentData(result), faceCapture.second),
                this::showError
        ));
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_SCAN_ID_CARD && resultCode == RESULT_OK && data != null) {
            viewBinding.progressBar.setVisibility(View.VISIBLE);
            viewBinding.mainUI.setVisibility(View.INVISIBLE);
            recognizerBundle.loadFromIntent(data);
            Recognizer<?> firstRecognizer = recognizerBundle.getRecognizers()[0];
            if (firstRecognizer instanceof BlinkIdCombinedRecognizer) {
                BlinkIdCombinedRecognizer.Result result = ((BlinkIdCombinedRecognizer)firstRecognizer).getResult();
                if (result.getDocumentDataMatch() == DataMatchResult.Failed) {
                    new AlertDialog.Builder(this)
                            .setTitle(R.string.invalid_licence)
                            .setMessage(R.string.front_and_back_dont_match)
                            .setNegativeButton(android.R.string.cancel, null)
                            .setPositiveButton(R.string.proceed_anyway, (dialogInterface, i) -> showCardFromResult(result))
                            .create()
                            .show();
                } else {
                    showCardFromResult(result);
                }
            }
        }
    }

    private void showSupportedDocuments() {
        Intent intent = new Intent(this, WebViewActivity.class);
        intent.putExtra(Intent.EXTRA_TEXT, BuildConfig.BLINK_SUPPORTED_DOCUMENTS_URL);
        startActivity(intent);
    }

    private void scanIDCard() {
        BlinkIdCombinedRecognizer recognizer = new BlinkIdCombinedRecognizer();
        recognizer.setReturnFullDocumentImage(true);
        recognizer.setEncodeFullDocumentImage(true);
        recognizer.setReturnFaceImage(true);
        recognizer.setEncodeFaceImage(true);
        recognizerBundle = new RecognizerBundle(recognizer);
        BlinkIdUISettings uiSettings = new BlinkIdUISettings(recognizerBundle);
        uiSettings.enableHighResSuccessFrameCapture(true);
        ActivityRunner.startActivityForResult(this, REQUEST_CODE_SCAN_ID_CARD, uiSettings);
    }

    private void showError(Throwable error) {
        viewBinding.progressBar.setVisibility(View.GONE);
        viewBinding.mainUI.setVisibility(View.VISIBLE);
        String message = error.getLocalizedMessage();
        if (error.getCause() != null && error.getCause().getLocalizedMessage() != null && !error.getCause().getLocalizedMessage().isEmpty()) {
            message += ": "+error.getCause().getLocalizedMessage();
        }
        new AlertDialog.Builder(this)
                .setTitle(R.string.error)
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, null)
                .create()
                .show();
    }

    private void showCard(FaceWithImage faceWithImage, DocumentData documentData, Float frontBackMatchScore) {
        viewBinding.progressBar.setVisibility(View.GONE);
        viewBinding.mainUI.setVisibility(View.VISIBLE);
        try {
            getSharedData().setSharedObject(IDCardActivity.EXTRA_FACE_IMAGE, faceWithImage);
            getSharedData().setSharedObject(IDCardActivity.EXTRA_DOCUMENT_DATA, documentData);
            Intent intent = new Intent(this, IDCardActivity.class);
            if (frontBackMatchScore != null) {
                intent.putExtra(IDCardActivity.EXTRA_FRONT_BACK_MATCH_SCORE, frontBackMatchScore);
            }
            startActivity(intent);
        } catch (Exception e) {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.error)
                    .setMessage("Failed to save captured ID card")
                    .setPositiveButton(android.R.string.ok, null)
                    .create()
                    .show();
        }
    }

    private void showAbout() {
        startActivity(new Intent(this, AboutActivity.class));
    }

    private void showSettings() {
        startActivity(new Intent(this, SettingsActivity.class));
    }
}

File: app/src/main/java/com/appliedrec/credentials/app/MainActivity.kt
package com.appliedrec.credentials.app

import android.annotation.SuppressLint
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.appliedrec.credentials.app.databinding.ActivityMainBinding
import com.appliedrec.verid.core2.*
import com.appliedrec.verid.core2.session.FaceCapture
import com.microblink.blinkid.entities.recognizers.Recognizer
import com.microblink.blinkid.entities.recognizers.RecognizerBundle
import com.microblink.blinkid.entities.recognizers.blinkid.generic.BlinkIdMultiSideRecognizer
import com.microblink.blinkid.entities.recognizers.blinkid.generic.classinfo.Region
import com.microblink.blinkid.entities.recognizers.blinkid.generic.classinfo.Type
import com.microblink.blinkid.uisettings.BlinkIdUISettings
import com.microblink.documentverification.client.DocumentVerificationClient
import com.microblink.documentverification.client.Response
import com.microblink.documentverification.client.data.model.request.DocumentVerificationRequest
import com.microblink.documentverification.client.data.model.request.ImageSource
import com.microblink.documentverification.client.data.model.result.DocumentVerificationResult
import com.microblink.documentverification.client.settings.DocumentVerificationServiceSettings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class MainActivity : AppCompatActivity() {

    private var blinkIdRecognizer: BlinkIdMultiSideRecognizer? = null
    private var blinkIdRecognizerBundle: RecognizerBundle? = null
    private var docVerClient: DocumentVerificationClient? = null
    private lateinit var viewBinding: ActivityMainBinding
    private lateinit var preferences: SharedPreferences
    private var capturedDocument: CapturedDocument? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        viewBinding.progressBar.visibility = View.VISIBLE
        viewBinding.content.visibility = View.GONE
        viewBinding.scanIDCardButton.setOnClickListener {
            captureDocument()
        }
        assert(application is IDCaptureApplication)
        (application as IDCaptureApplication).verIDLiveData.observe(this) { verIDResult ->
            if (verIDResult == null) {
                return@observe
            }
            viewBinding.content.visibility = View.VISIBLE
            viewBinding.progressBar.visibility = View.GONE
            if (verIDResult.isSuccess) {
                observeDocumentCapture()
            } else {
                val error = verIDResult.exceptionOrNull()?.localizedMessage ?: "Not sure why"
                showError("Failed to load Ver-ID", error)
            }
        }
        viewBinding.settings.setOnClickListener {
            startActivity(Intent(this, SettingsActivity::class.java))
        }
    }

    private fun observeDocumentCapture() {
        (application as IDCaptureApplication).documentLiveData.observe(this) { result ->
            if (result == null) {
                return@observe
            }
            viewBinding.content.visibility = View.VISIBLE
            viewBinding.progressBar.visibility = View.GONE
            if (result.isSuccess) {
                startActivity(Intent(this, IDDocumentActivity::class.java))
            } else {
                val error = result.exceptionOrNull()?.localizedMessage ?: "Something went wrong"
                showError("Scan failed", error)
            }
        }
    }

    private fun createDocumentVerificationClient(): DocumentVerificationClient {
        val metaData = getAppMetadata()
        val docVerURL = metaData.getString("docver.url") ?: throw Exception()
        val docVerClientId = metaData.getString("docver.clientid") ?: throw Exception()
        val docVerClientSecret = metaData.getString("docver.clientsecret") ?: throw Exception()
        val docVerClientSettings = DocumentVerificationServiceSettings(docVerURL, docVerClientId, docVerClientSecret)
        return DocumentVerificationClient(docVerClientSettings)
    }

    private fun getAppMetadata(): Bundle {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            this.packageManager.getApplicationInfo(this.packageName, PackageManager.ApplicationInfoFlags.of(
                PackageManager.GET_META_DATA.toLong()
            )).metaData
        } else {
            this.packageManager.getApplicationInfo(this.packageName, PackageManager.GET_META_DATA).metaData
        }
    }

    private suspend fun processDocumentCaptureResult(blinkIdResult: BlinkIdMultiSideRecognizer.Result): Pair<Bitmap,DocumentVerificationResult?> {
        val bitmap: Bitmap
        var docVerResult: DocumentVerificationResult? = null
        if (docVerClient != null) {
            val frontBitmap = blinkIdResult.frontCameraFrame.convertToBitmap() ?: throw Exception()
            val backBitmap = blinkIdResult.backCameraFrame.convertToBitmap() ?: throw Exception()
            val frontImage = ImageSource(frontBitmap)
            val backImage = ImageSource(backBitmap)
            val docVerificationRequest = DocumentVerificationRequest(frontImage, backImage, true)
            when (val response = docVerClient!!.verify(docVerificationRequest)) {
                is Response.Error -> {
                    throw response.exception ?: Exception()
                }
                is Response.Success -> {
                    bitmap = response.endpointResponse.data?.extractionResult?.fullDocumentFrontImage
                            ?: throw Exception()
                    docVerResult = response.endpointResponse.data
                }
            }
        } else {
            bitmap = blinkIdResult.fullDocumentFrontImage?.convertToBitmap() ?: throw Exception()
        }
        return Pair(bitmap, docVerResult)
    }

    private suspend fun getDocumentBackImage(blinkIdResult: BlinkIdMultiSideRecognizer.Result): Bitmap? {
        if (blinkIdRecognizer?.shouldSaveCameraFrames() == true) {
            return blinkIdResult.backCameraFrame.convertToBitmap()
        } else if (blinkIdRecognizer?.shouldReturnFullDocumentImage() == true) {
            return blinkIdResult.fullDocumentBackImage?.convertToBitmap()
        } else {
            return null
        }
    }

    private val barcodeScanLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == RESULT_OK && result.data?.hasExtra(Intent.EXTRA_TEXT) == true) {
            val barcode = result.data?.getByteArrayExtra(Intent.EXTRA_TEXT)
            if (barcode != null && capturedDocument != null) {
                capturedDocument!!.rawBarcode = barcode.decodeToString()
                (application as IDCaptureApplication).documentLiveData.postValue(
                    Result.success(capturedDocument!!)
                )
                capturedDocument = null
            } else {
                showError("Scan failed", "No barcode found")
            }
        } else if (result.resultCode == RESULT_CANCELED) {
            viewBinding.content.visibility = View.VISIBLE
            viewBinding.progressBar.visibility = View.GONE
        }
    }

    @SuppressLint("RestrictedApi")
    private val documentCaptureLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == RESULT_OK && result.data != null) {
            blinkIdRecognizerBundle!!.loadFromIntent(result.data!!)
            val blinkIdResult = blinkIdRecognizer!!.result
            if (blinkIdResult.resultState == Recognizer.Result.State.Valid) {
                lifecycleScope.launch(Dispatchers.IO) {
                    try {
                        val bitmapAndDocVerResult = processDocumentCaptureResult(blinkIdResult)
                        var rawBarcode: String? = null
                        if (blinkIdResult.classInfo.type == Type.HEALTH_INSURANCE_CARD && blinkIdResult.classInfo.region == Region.ONTARIO) {
                            getDocumentBackImage(blinkIdResult)?.let { backImage ->
                                try {
                                    rawBarcode = BarcodeDetector.detectBarcodeInImage(backImage)
                                } catch (error: Exception) {
                                    Log.e("BarcodeDetector", "Failed to detect barcode", error)
                                }
                            }
                        }
                        val image = Image(bitmapAndDocVerResult.first)
                        val detectAuthenticity: Boolean = AuthenticityScoreSupport.isDocumentSupported(blinkIdResult)
                        val faceAuthenticityPair = detectFaceInImage(image, detectAuthenticity)
                        val faceCapture = FaceCapture(
                            faceAuthenticityPair.first,
                            Bearing.STRAIGHT,
                            bitmapAndDocVerResult.first,
                            image
                        )
                        val document = CapturedDocument(
                            documentCaptureResult = blinkIdResult,
                            faceCapture = faceCapture,
                            documentVerificationResult = bitmapAndDocVerResult.second,
                            authenticityScore = faceAuthenticityPair.second,
                            rawBarcode = rawBarcode
                        )
                        if (document.type == Type.HEALTH_INSURANCE_CARD && document.region == Region.ONTARIO && rawBarcode == null) {
                            withContext(Dispatchers.Main) {
                                capturedDocument = document
                                captureOntarioHealthCardBarcode()
                            }
                        } else {
                            (application as IDCaptureApplication).documentLiveData.postValue(
                                Result.success(document)
                            )
                        }
                    } catch (error: Exception) {
                        (application as IDCaptureApplication).documentLiveData.postValue(Result.failure(error))
                    }
                }
            }
        } else if (result.resultCode == RESULT_CANCELED) {
            viewBinding.content.visibility = View.VISIBLE
            viewBinding.progressBar.visibility = View.GONE
        }
    }

    private suspend fun detectFaceInImage(image: Image, detectAuthenticity: Boolean): Pair<RecognizableFace,Float?> = suspendCoroutine { continuation ->
        val verID = (application as IDCaptureApplication).verIDLiveData.value!!.getOrElse {
            continuation.resumeWithException(it)
            return@suspendCoroutine
        }
        try {
            val faceDetectionUtilities = verID.utilities.map { it.faceDetectionUtilities }
                .orElseThrow { Exception("Face detection utilities not available") }
            faceDetectionUtilities.detectRecognizableFacesInImage(image, 1, EnumSet.of(VerIDFaceTemplateVersion.getLatest())) { faceDetectionResult ->
                try {
                    var authenticityScore: Float? = null
                    val face = faceDetectionResult.get().firstOrNull() ?: throw Exception("No face detected")
                    if (detectAuthenticity && verID.faceDetection is FaceDetection) {
                        val classifier = AuthenticityScoreSupport.getClassifiers(this).firstOrNull()
                        if (classifier != null) {
                            val faceDetection = verID.faceDetection as FaceDetection
                            authenticityScore = faceDetection.extractAttributeFromFace(face, image, classifier.name)
                        }
                    }
                    continuation.resume(Pair(face, authenticityScore))
                } catch (e: Exception) {
                    continuation.resumeWithException(e)
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    private fun captureOntarioHealthCardBarcode() {
        AlertDialog.Builder(this)
            .setTitle("Barcode scan failed")
            .setMessage("We were unable to read the barcode on the back of the card")
            .setPositiveButton("Retry") { _, _ ->
                barcodeScanLauncher.launch(Intent(this, BarcodeScannerActivity::class.java))
            }
            .setNegativeButton("Cancel", null)
            .create()
            .show()
    }

    private fun captureDocument() {
        viewBinding.content.visibility = View.GONE
        viewBinding.progressBar.visibility = View.VISIBLE
        blinkIdRecognizer = BlinkIdMultiSideRecognizer()
        if (preferences.enableDocumentVerification) {
            blinkIdRecognizer!!.setSaveCameraFrames(true)
            docVerClient = createDocumentVerificationClient()
        } else {
            blinkIdRecognizer!!.setReturnFullDocumentImage(true)
            docVerClient = null
        }
        blinkIdRecognizerBundle = RecognizerBundle(blinkIdRecognizer)
        val settings = BlinkIdUISettings(blinkIdRecognizerBundle)
        val intent = Intent(this, settings.targetActivity)
        settings.saveToIntent(intent)
        documentCaptureLauncher.launch(intent)
    }
}